linera-client/src/benchmark.rs

```diff
@@ -170,10 +287,196 @@ where
+                )?,
+                &histogram,
+            )?;
+            let p99 = match Self::compute_quantile(&diff.buckets, diff.count, 0.99) {
+                Ok(p99) => p99,
+                Err(BenchmarkError::NoDataYetForP99Calculation) => {
+                    info!(
+                        "No data available yet to calculate p99 for {}",
+                        metrics_address
+                    );
+                    continue;
+                }
+                Err(e) => {
+                    error!("Error computing p99 for {}: {}", metrics_address, e);
+                    return Err(e);
+                }
+            };
+
+            let last_bucket_boundary = diff.buckets[diff.buckets.len() - 2].0;
+            if p99 == f64::INFINITY {
+                info!(
+                    "{} -> Estimated p99 for {} is higher than the last bucket boundary of {:?} ms",
+                    metrics_address, LATENCY_METRIC_PREFIX, last_bucket_boundary
+                );
+            } else {
+                info!(
+                    "{} -> Estimated p99 for {}: {:.2} ms",
+                    metrics_address, LATENCY_METRIC_PREFIX, p99
+                );
+            }
+            if p99 > PROXY_LATENCY_P99_THRESHOLD {
+                if p99 == f64::INFINITY {
+                    error!(
+                        "Proxy of validator {} unhealthy! Latency p99 is too high, it is higher than the last bucket boundary of {:.2} ms",
+                        metrics_address, last_bucket_boundary
+                    );
+                } else {
+                    error!(
+                        "Proxy of validator {} unhealthy! Latency p99 is too high: {:.2} ms",
+                        metrics_address, p99
+                    );
+                }
+                return Ok(false);
+            }
+            previous_histogram_snapshots.insert(metrics_address.clone(), histogram);
+        }
+
+        Ok(true)
+    }
+
+    fn diff_histograms(
+        previous: &HistogramSnapshot,
+        current: &HistogramSnapshot,
+    ) -> Result<HistogramSnapshot, BenchmarkError> {
+        if current.count < previous.count {
+            return Err(BenchmarkError::HistogramCountMismatch);
+        }
+        let total_diff = current.count - previous.count;
+        let mut buckets_diff: Vec<(f64, f64)> = Vec::new();
+        for ((bound_before, cumulative_before), (bound_after, cumulative_after)) in
+            previous.buckets.iter().zip(current.buckets.iter())
+        {
+            if (bound_before - bound_after).abs() > f64::EPSILON {
+                return Err(BenchmarkError::BucketBoundariesDoNotMatch(
+                    *bound_before,
+                    *bound_after,
+                ));
+            }
+            let diff = if cumulative_after >= cumulative_before {
+                cumulative_after - cumulative_before
+            } else {
+                0.0
+            };
```

Comment:
(Maybe more concise.)

```suggestion
            let diff = (cumulative_after - cumulative_before).max(0.0);
```

Reply:
True 😅